application/forms/EventRuleConfigForm.php

```diff
@@ -0,0 +1,654 @@
+                }
+            }
+
+            foreach ($recipientsFromConfig as $recipientConfig) {
+                $data = [
+                    'rule_escalation_id' => $escalationId,
+                    'channel_id'         => $recipientConfig['channel_id']
+                ];
+
+                switch (true) {
+                    case isset($recipientConfig['contact_id']):
+                        $data['contact_id'] = $recipientConfig['contact_id'];
+                        $data['contactgroup_id'] = null;
+                        $data['schedule_id'] = null;
+                        break;
+                    case isset($recipientConfig['contactgroup_id']):
+                        $data['contact_id'] = null;
+                        $data['contactgroup_id'] = $recipientConfig['contactgroup_id'];
+                        $data['schedule_id'] = null;
+                        break;
+                    case isset($recipientConfig['schedule_id']):
+                        $data['contact_id'] = null;
+                        $data['contactgroup_id'] = null;
+                        $data['schedule_id'] = $recipientConfig['schedule_id'];
+                        break;
+                }
+
+                if (! isset($recipientConfig['id'])) {
+                    $db->insert('rule_escalation_recipient', $data);
+                } else {
+                    $db->update('rule_escalation_recipient', $data, ['id = ?' => $recipientConfig['id']]);
+                }
+            }
+        }
+    }
+
+    public function isValidEvent($event)
+    {
+        if (in_array($event, [self::ON_CHANGE, self::ON_DELETE, self::ON_DISCARD])) {
+            return true;
+        }
+
+        return parent::isValidEvent($event);
+    }
+
+    /**
+     * Remove the given event rule
+     *
+     * @param int $id
+     *
+     * @return void
+     */
+    public function removeRule(int $id): void
+    {
+        $db = Database::get();
+        $db->beginTransaction();
+        $escalations = RuleEscalation::on($db)
+            ->columns('id')
+            ->filter(Filter::equal('rule_id', $id));
+
+        $escalationsToRemove = [];
+        /** @var RuleEscalation $escalation */
+        foreach ($escalations as $escalation) {
+            $escalationsToRemove[] = $escalation->id;
+        }
+
+        if (! empty($escalationsToRemove)) {
+            $db->delete('rule_escalation_recipient', ['rule_escalation_id IN (?)' => $escalationsToRemove]);
+        }
+
+        $db->delete('rule_escalation', ['rule_id = ?' => $id]);
+        $db->delete('rule', ['id = ?' => $id]);
+
+        $db->commitTransaction();
+    }
+
+    /**
+     * Get the prefix map
+     *
+     * @param int $escalationCount
+     *
+     * @return string
+     */
+    protected function getPrefixesMap(int $escalationCount): string
+    {
+        $prefixesMap = [];
+        for ($i = 1; $i <= $escalationCount; $i++) {
+            $prefixesMap[(string) $i] = bin2hex((string) $i);
```

Comment:
Why can't the `key` be omitted?

Reply:
It is required to populate and remove right escalations.

Reply:
I can't find a use for it. Can you please show me the lines of code.